src/main/resources/jupyter/init-actions.sh

```diff
@@ -160,130 +164,142 @@ if [[ "${ROLE}" == 'Master' ]]; then
+      COMPOSE_FILES+=(-f /etc/`basename ${JUPYTER_DOCKER_COMPOSE}`)
     fi
-
-    #Install combined extensions
-    if [ ! -z "${JUPYTER_COMBINED_EXTENSIONS}"  ] ; then
-      for ext in ${JUPYTER_COMBINED_EXTENSIONS}
-      do
-        log 'Installing Jupyter combined extension [$ext]...'
-        log $ext
-        if [[ $ext == 'gs://'* ]]; then
-          gsutil cp $ext /etc
-          JUPYTER_EXTENSION_ARCHIVE=`basename $ext`
-          docker cp /etc/${JUPYTER_EXTENSION_ARCHIVE} ${JUPYTER_SERVER_NAME}:${JUPYTER_HOME}/${JUPYTER_EXTENSION_ARCHIVE}
-          retry 3 docker exec -u root -e PIP_USER=false ${JUPYTER_SERVER_NAME} ${JUPYTER_SCRIPTS}/extension/jupyter_install_combined_extension.sh ${JUPYTER_EXTENSION_ARCHIVE}
-        else
-          retry 3 docker exec -u root -e PIP_USER=false ${JUPYTER_SERVER_NAME} ${JUPYTER_SCRIPTS}/extension/jupyter_pip_install_combined_extension.sh $ext
-        fi
-      done
+    if [ ! -z ${RSTUDIO_DOCKER_IMAGE} ] ; then
+      COMPOSE_FILES+=(-f /etc/`basename ${RSTUDIO_DOCKER_COMPOSE}`)
     fi
-
-    # If a custom.js was specified, copy it into the jupyter docker container.
-    if [ ! -z ${JUPYTER_CUSTOM_JS_URI} ] ; then
-      log 'Installing Jupyter custom.js...'
-      gsutil cp ${JUPYTER_CUSTOM_JS_URI} /etc
-      JUPYTER_CUSTOM_JS=`basename ${JUPYTER_CUSTOM_JS_URI}`
-      retry 3 docker exec ${JUPYTER_SERVER_NAME} mkdir -p ${JUPYTER_USER_HOME}/.jupyter/custom
-      docker cp /etc/${JUPYTER_CUSTOM_JS} ${JUPYTER_SERVER_NAME}:${JUPYTER_USER_HOME}/.jupyter/custom/
-    fi
-
-    # If a google_sign_in.js was specified, copy it into the jupyter docker container.
-    if [ ! -z ${JUPYTER_GOOGLE_SIGN_IN_JS_URI} ] ; then
-      log 'Installing Google sign in extension...'
-      gsutil cp ${JUPYTER_GOOGLE_SIGN_IN_JS_URI} /etc
-      JUPYTER_GOOGLE_SIGN_IN_JS=`basename ${JUPYTER_GOOGLE_SIGN_IN_JS_URI}`
-      retry 3 docker exec ${JUPYTER_SERVER_NAME} mkdir -p ${JUPYTER_USER_HOME}/.jupyter/custom
-      docker cp /etc/${JUPYTER_GOOGLE_SIGN_IN_JS} ${JUPYTER_SERVER_NAME}:${JUPYTER_USER_HOME}/.jupyter/custom/
+    docker-compose "${COMPOSE_FILES[@]}" config
+    retry 5 docker-compose "${COMPOSE_FILES[@]}" pull
+    docker-compose "${COMPOSE_FILES[@]}" up -d
+
+    # Jupyter-specific setup, only do if Jupyter is installed
+    if [ ! -z ${JUPYTER_DOCKER_IMAGE} ] ; then
+      log 'Installing Jupydocker kernelspecs...'
+
+      # Change Python and PySpark 2 and 3 kernel specs to allow each to have its own spark
+      retry 3 docker exec -u root ${JUPYTER_SERVER_NAME} ${JUPYTER_SCRIPTS}/kernel/kernelspec.sh ${JUPYTER_SCRIPTS}/kernel ${KERNELSPEC_HOME}
+
+      log 'Installing Hail additions to Jupydocker spark.conf...'
+
+      # Install the Hail additions to Spark conf.
+      retry 3 docker exec -u root ${JUPYTER_SERVER_NAME} ${JUPYTER_SCRIPTS}/hail/spark_install_hail.sh
+
+      # Copy the actual service account JSON file into the Jupyter docker container.
+      # TODO: not Jupyter specific
```

Comment:
I think I had noted this in my PR, but would be good to lift this out if the `if [ ! -z ${JUPYTER_DOCKER_IMAGE} ] ; then` block.